torch/distributed/algorithms/quantization.py

```diff
@@ -0,0 +1,141 @@
+import functools
+import torch
+# import torch.autograd
+import torch.distributed as dist
+# import torch.nn
+
+
+from enum import Enum
+
+
+TORCH_HALF_MIN = torch.finfo(torch.float16).min
+TORCH_HALF_MAX = torch.finfo(torch.float16).max
+
+class QuantType(Enum):
```

Comment:
we have QuantType defined here as well: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quant_type.py to reduce confusion, should we rename this to something like DQuantType (meaning distributed quant type)? or if you only want to do float16/bfloat16 I think it would be good enough to just pass around dtype directly instead of creating a new type here.

Reply:
The reason we are creating the new Enum class and the qtype is that we want to support other quantization methods in the future. One of my stretch goals is to add INT8 type.

Reply:
I think even for int8 type, we can do that with dtype probably. will you have different ways to quantize for the same dtype?

Also are we just quantizing weights? or activation as well?

Reply:
> I think even for int8 type, we can do that with dtype probably. will you have different ways to quantize for the same dtype?

yeah we can do `dtype` directly, but as you said, we do have plans to support all quantization types that `torch.quantization` offers, including BFP16 and different quantization schemes, so this type might expand to including both `dtype` and [`torch.qscheme`](https://pytorch.org/docs/stable/quantization-support.html?highlight=qscheme) to let user specify in details how to quantize the tensor before the collective call, so exposed a class here as user's API here might be good.  Do you think this is legit or have any suggestions on supporting both quantization dtype/scheme as we evolve?

> Also are we just quantizing weights? or activation as well?

I think no weights or activation concepts here, we are just trying to quantize/dequantize tensors before/after the collective, so any tensor/tensor list passed to the collective call, we will try to quantize/dequantize them?

Reply:
I see, there are some quantization types like static quantization that may not make sense here I feel, but we can discuss later. I think the only thing that makes sense for int8 is dynamically quantize the Tensor (that is, we calculate min/max on the single Tensor that we want to quantize and then quantize it), that's why I feel dtype is good enough. qscheme is probably a separate discussion, I think it makes sense to add qscheme in the future.